### DISCUSSION TITLE
Is there a way of creating a custom provider without relying on Terraform or Pulumi?

### DISCUSSION THREAD

**Discussion by brunocascio** [nitrictech/nitric#708] [Q&A] [answered]
I'm wondering if it's possible to create a custom provider without relying on terraform/pulumi and if there's some documentation about it.

For example, I would like to create a provider based on an REST API without the need of creating Terraform/Pulumi providers first.

Documentation shows the use of protobuf for it, but I didn't find docs about it.

Is it possible?

**Comment by jyecusch** [comment, marked as answer]
Hi @brunocascio, it's defintely possible to build a custom provider without relying on Terraform/Pulumi. However, this hasn't been a common request, so the documentation is a bit limited.

Here is the current docs page about fully custom providers https://nitric.io/docs/providers/custom/create

Essentially a Nitric provider implements two gRPC servers:
- a deployment server which is responsible for translating resource requirements into some way of deploying the resources
- a runtime server which acts as an adapter for the target platform

These two implementations can be written in any language that's supported by gRPC/Protocol Buffers. All of standard the providers we offer are written in Go. We chose Go because it had the best performance/compatibility balance (e.g. Rust might be faster, but not all clouds have Rust SDKs).

Here is an example of a partially implemented Provider in Python https://github.com/nitrictech/python-terraform-provider-poc, in case that's helpful.

I'd be happy to answer other questions and if the docs can be improved to help you we'll try to get that done too.

**Reply by brunocascio** [reply]
Thanks a lot @jyecusch! 
I wish luck you guys on this, really like this product 🙌
